### DISCUSSION TITLE
How to check the type of a js value, such as whether it is undefined

### DISCUSSION THREAD

**Discussion by 2A5F** [ClearFoundry/ClearScript#513] [Q&A] [answered]
Now it can only be converted into a string to check

```cs
var some = rt.Evaluate("");
some.ToString() == "[undefined]"
```
Can you provide an API that returns JsValue instead of just dynamic

**Comment by 2A5F** [comment]
Well, I found that some == Undefined.Value, but other types seem to be unable to check

**Comment by ClearScriptLib** [comment, marked as answer]
Hello @2A5F,

> Can you provide an API that returns JsValue instead of just dynamic

ClearScript doesn't have such APIs because (a) scripts can return .NET values as well as script values, and (b) ClearScript's API is designed to support languages other than JavaScript.

You can use this summary to identify and interact with data returned from script code:

- All .NET values are returned in their native .NET form.
- Script numeric, Boolean, and string values are converted to their equivalent .NET types.
- JavaScript [`undefined`](https://developer.mozilla.org/en-US/docs/Glossary/Undefined) is converted to [`Undefined.Value`](https://clearscript.clearfoundry.net/Reference/html/F_Microsoft_ClearScript_Undefined_Value.htm).
- All script objects implement [`IScriptObject`](https://clearscript.clearfoundry.net/Reference/html/T_Microsoft_ClearScript_IScriptObject.htm).
- [V8] JavaScript [`BigInt`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt) is converted to [`BigInteger`](https://learn.microsoft.com/en-us/dotnet/api/system.numerics.biginteger).
- [V8] All JavaScript objects implement [`IJavaScriptObject`](https://clearscript.clearfoundry.net/Reference/html/T_Microsoft_ClearScript_JavaScript_IJavaScriptObject.htm), which provides additional type information.
- [V8] Generic JavaScript objects implement [`IDictionary<string, object>`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.idictionary-2).
- [V8] JavaScript arrays implement [`IList`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.ilist) and [`IList<object>`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.ilist-1).
- [V8] JavaScript [array buffers](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer) implement [`IArrayBuffer`](https://clearscript.clearfoundry.net/Reference/html/T_Microsoft_ClearScript_JavaScript_IArrayBuffer.htm).
- [V8] JavaScript [data views](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView) implement [`IDataView`](https://clearscript.clearfoundry.net/Reference/html/T_Microsoft_ClearScript_JavaScript_IDataView.htm).
- [V8] JavaScript [typed arrays](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Typed_arrays) implement [`ITypedArray<T>`](https://clearscript.clearfoundry.net/Reference/html/T_Microsoft_ClearScript_JavaScript_ITypedArray_1.htm).

Good luck!
